File: domain/stats/stats.go
/*
	能力値を計算してstats.Statsを生成する
*/
package stats

import (
	"damagecalculator/domain/basepoints"
	"damagecalculator/domain/individuals"
)

type stats struct {
	HP, Attack, Defense, SpAttack, SpDefense, Speed uint
}

func CalcHP(l Level, s *SpeciesStats, i individuals.Individuals, b uint, n *Nature) uint {
	return n.calcHP(l, s.hp, i.HP(), basePoints.NewBasePoint(b))
}
func CalcAttack(l Level, s *SpeciesStats, i individuals.Individuals, b uint, n *Nature) uint {
	return n.calcAttack(l, s.at, i.Attack(), basePoints.NewBasePoint(b))
}
func CalcDefense(l Level, s *SpeciesStats, i individuals.Individuals, b uint, n *Nature) uint {
	return n.calcDefense(l, s.df, i.Defense(), basePoints.NewBasePoint(b))
}
func CalcSpAttack(l Level, s *SpeciesStats, i individuals.Individuals, b uint, n *Nature) uint {
	return n.calcSpAttack(l, s.sa, i.SpAttack(), basePoints.NewBasePoint(b))
}
func CalcSpDefense(l Level, s *SpeciesStats, i individuals.Individuals, b uint, n *Nature) uint {
	return n.calcSpDefense(l, s.sd, i.SpDefense(), basePoints.NewBasePoint(b))
}
func CalcSpeed(l Level, s *SpeciesStats, i individuals.Individuals, b uint, n *Nature) uint {
	return n.calcSpeed(l, s.sp, i.Speed(), basePoints.NewBasePoint(b))
}

func NewStats(l Level, s *SpeciesStats, i individuals.Individuals, b basePoints.BasePoints, n *Nature) *stats {
	hp := n.calcHP(l, s.hp, i.HP(), b.HP())
	at := n.calcAttack(l, s.at, i.Attack(), b.Attack())
	df := n.calcDefense(l, s.df, i.Defense(), b.Defense())
	sa := n.calcSpAttack(l, s.sa, i.SpAttack(), b.SpAttack())
	sd := n.calcSpDefense(l, s.sd, i.SpDefense(), b.SpDefense())
	sp := n.calcSpeed(l, s.sp, i.Speed(), b.Speed())

	return &stats{
		HP:        hp,
		Attack:    at,
		Defense:   df,
		SpAttack:  sa,
		SpDefense: sd,
		Speed:     sp,
	}
}

File: frontend/src/domain/stats.ts
import { INature } from './nature';

export class StatsCalculator {
  constructor() {
  }
  Calculate(nature: INature, level: number, species: number[], individuals: number[], basePoints: number[]): number[] {
    let hp = this.hp(level, species[0], individuals[0], basePoints[0]);
    let at = nature.Attack(this.stats(level, species[1], individuals[1], basePoints[1]));
    let df = nature.Defense(this.stats(level, species[2], individuals[2], basePoints[2]));
    let sa = nature.SpAttack(this.stats(level, species[3], individuals[3], basePoints[3]));
    let sd = nature.SpDefense(this.stats(level, species[4], individuals[4], basePoints[4]));
    let sp = nature.Speed(this.stats(level, species[5], individuals[5], basePoints[5]));

    let res: number[] = [hp, at, df, sa, sd, sp];
    return res;
  }

  private hp(l: number, s: number, i: number, b: number): number {
    let x = s * 2 + i + b / 4;
    let y = x * l / 100.0 + l + 10;
    return Math.floor(y);
  }
  private stats(l: number, s: number, i: number, b: number): number {
    let x = s * 2 + i + b / 4;
    let y = Math.floor(x * l / 100.0 + 5);
    return Math.floor(y);
  }
}
